File: backend/app/validator.py
import pandas as pd
import io
import re
from typing import List
from .schemas import ValidationError, ValidationResponse

REQUIRED_COLUMNS = ["id", "email", "age"]

def validate_csv(file_content: bytes) -> ValidationResponse:
    try:
        df = pd.read_csv(io.BytesIO(file_content))
    except Exception as e:
        return ValidationResponse(
            status="fail",
            errors=[ValidationError(error_message=f"Failed to parse CSV: {str(e)}")]
        )

    errors: List[ValidationError] = []

    # Check for missing columns
    missing_cols = [col for col in REQUIRED_COLUMNS if col not in df.columns]
    if missing_cols:
        return ValidationResponse(
            status="fail",
            errors=[ValidationError(error_message=f"Missing required columns: {', '.join(missing_cols)}")]
        )

    # Check for minimum row count
    if len(df) <= 10:
        return ValidationResponse(
            status="fail",
            errors=[ValidationError(row_index=None, id=None, column=None, error_message="File contains 10 or fewer data rows.")]
        )

    # --- Vectorized Validation ---
    
    # 1. Email validation (empty or null)
    email_invalid_mask = df['email'].isna() | (df['email'].astype(str).str.strip() == "")
    email_errors_df = df[email_invalid_mask]
    for idx, row in email_errors_df.iterrows():
        errors.append(ValidationError(
            row_index=int(idx) + 1,
            id=row.get("id"),
            column="email",
            error_message="Email column must not be empty or null."
        ))

    # 2. Age validation
    # Convert 'age' to numeric, invalid entries become NaN
    numeric_age = pd.to_numeric(df['age'], errors='coerce')
    
    # 2a. Identify rows with invalid number format (wasn't NaN but became NaN)
    format_errors_mask = numeric_age.isna()
    format_errors_df = df[format_errors_mask]
    for idx, row in format_errors_df.iterrows():
        errors.append(ValidationError(
            row_index=int(idx) + 1,
            id=row.get("id"),
            column="age",
            error_message=f"Invalid number format: '{row.get('age')}'"
        ))
        
    # 2b. Range errors (only for valid numeric entries)
    range_errors_mask = ~numeric_age.isna() & ((numeric_age < 18) | (numeric_age > 100))
    range_errors_df = df[range_errors_mask]
    for idx, row in range_errors_df.iterrows():
        # Using .loc to get the numeric age for the current index
        age_val = int(numeric_age.loc[idx])
        errors.append(ValidationError(
            row_index=int(idx) + 1,
            id=row.get("id"),
            column="age",
            error_message=f"Age {age_val} is outside the allowed range of 18-100."
        ))

    status = "pass" if not errors else "fail"
    return ValidationResponse(status=status, errors=errors)

File: rust-backend/src/validator.rs
use csv::ReaderBuilder;
use crate::schemas::{ValidationError, ValidationResponse};
use std::io::Cursor;
use bytes::Bytes;

const REQUIRED_COLUMNS: &[&str] = &["id", "email", "age"];

pub fn validate_csv(content: Bytes) -> ValidationResponse {
    let cursor = Cursor::new(content);
    let mut rdr = ReaderBuilder::new()
        .has_headers(true)
        .from_reader(cursor);

    // 1. Check headers
    let headers = match rdr.headers() {
        Ok(h) => h.clone(),
        Err(e) => return fail(format!("Failed to parse CSV: {}", e)),
    };
    
    let header_names: Vec<&str> = headers.iter().collect();
    let mut missing_cols = Vec::new();
    for req in REQUIRED_COLUMNS {
        if !header_names.contains(req) {
            missing_cols.push(*req);
        }
    }
    
    if !missing_cols.is_empty() {
        return fail(format!("Missing required columns: {}", missing_cols.join(", ")));
    }
    
    let mut records = Vec::new();
    for result in rdr.records() {
        match result {
            Ok(rec) => records.push(rec),
            Err(e) => return fail(format!("Failed to read record: {}", e)),
        }
    }
    
    // 2. Check row count
    // Python code checks: if len(df) <= 10
    if records.len() <= 10 {
        return ValidationResponse {
            status: "fail".to_string(),
            errors: vec![ValidationError {
                row_index: None,
                id: None,
                column: None,
                error_message: "File contains 10 or fewer data rows.".to_string(),
            }],
        };
    }
    
    let mut errors = Vec::new();
    
    let id_idx = headers.iter().position(|h| h == "id").unwrap();
    let email_idx = headers.iter().position(|h| h == "email").unwrap();
    let age_idx = headers.iter().position(|h| h == "age").unwrap();
    
    for (i, record) in records.iter().enumerate() {
        // Python uses 1-based indexing for rows
        let row_index = i + 1;
        let id_val = record.get(id_idx).map(|s| s.to_string());
        
        // 1. Email validation
        let email = record.get(email_idx).unwrap_or("");
        if email.trim().is_empty() {
             errors.push(ValidationError {
                row_index: Some(row_index),
                id: id_val.clone(),
                column: Some("email".to_string()),
                error_message: "Email column must not be empty or null.".to_string(),
            });
        }
        
        // 2. Age validation
        let age_str = record.get(age_idx).unwrap_or("");
        
        // Try to parse as float first (to match Pandas numeric behavior usually handling floats)
        // If it can be parsed, we check range. If not, it's a format error.
        match age_str.parse::<f64>() {
             Ok(age_num) => {
                 // Check range 18-100
                 if age_num < 18.0 || age_num > 100.0 {
                     errors.push(ValidationError {
                        row_index: Some(row_index),
                        id: id_val.clone(),
                        column: Some("age".to_string()),
                        error_message: format!("Age {} is outside the allowed range of 18-100.", age_num as i64),
                    });
                 }
             }
             Err(_) => {
                 // Invalid number format
                 errors.push(ValidationError {
                    row_index: Some(row_index),
                    id: id_val.clone(),
                    column: Some("age".to_string()),
                    error_message: format!("Invalid number format: '{}'", age_str),
                });
             }
        }
    }
    
    let status = if errors.is_empty() { "pass" } else { "fail" };
    ValidationResponse {
        status: status.to_string(),
        errors,
    }
}

fn fail(msg: String) -> ValidationResponse {
    ValidationResponse {
        status: "fail".to_string(),
        errors: vec![ValidationError {
            row_index: None,
            id: None,
            column: None,
            error_message: msg,
        }],
    }
}
